File: 07.CodeAct/codeact/graph.py
from langchain_core.messages import AIMessage, SystemMessage,HumanMessage, ToolMessage
from langgraph.types import Send
from langgraph.graph import StateGraph, MessagesState
from langgraph.graph import START, END
from langgraph.graph.message import add_messages
from llm import Tongyi
from prompts import SYSTEM_PROMPT
from typing import Annotated, TypedDict
from tools import tools

llm_Tongyi = Tongyi()

class OverallState(TypedDict):
    messages: Annotated[list, add_messages]
    actions: list[dict]
    output: str
    user_prompt: str

class CodeActGraph():
    def run(self, user_prompt: str):
        def _create_agent_action(ai_message):
            # 提取代码部分
            content = ai_message.content
            if "```python" in content:
                code_blocks = content.split("```python")
                code = code_blocks[1].split("```")[0].strip()
                return {"tool": "execute_python", "tool_input": code}
            return None

        def llm_call(state: OverallState):
            # 构建包含系统提示词和用户提示词的消息列表
            messages = [
                SystemMessage(content=SYSTEM_PROMPT),
                HumanMessage(content=state["user_prompt"])
            ]+state["messages"]

            ret = llm_Tongyi.invoke(messages)

            action = _create_agent_action(ret)
            if action:
                return {"messages": [ret], "actions": [action], "output": ""}
            return {"messages": [ret], "output": ret.content}

        def enter_process(state: OverallState):
            if state["output"] != "":
                return END
            return "process_node"

        def process_node(state: OverallState):
            actions = state["actions"]
            for action in actions:
                tool_name = action["tool"]
                tool_input = action["tool_input"]
                tool_fn = next(t for t in tools if t.name == tool_name)
                observation = tool_fn.invoke(tool_input)  #执行工具
                state["messages"].append(HumanMessage(content=f"##执行结果:\n{observation}"))

            return state

        graph = StateGraph(OverallState)
        graph.add_node("llm_call", llm_call)
        graph.add_node("enter_process", enter_process)
        graph.add_node("process_node", process_node)

        graph.add_edge(START, "llm_call")
        graph.add_conditional_edges("llm_call",enter_process)
        graph.add_edge("process_node", "llm_call")

        agent = graph.compile()
        ret = agent.invoke(OverallState(user_prompt=user_prompt))
        
        return ret["output"]

if __name__ == "__main__":
    graph = CodeActGraph()
    ret = graph.run("请计算1~100的和")
    print(ret)

File: js/src/codeact/graph.ts
import {StateGraph, END} from '@langchain/langgraph'
import {ChatOpenAI} from '@langchain/openai'
import {HumanMessage, SystemMessage} from '@langchain/core/messages'
import {SYSTEM_PROMPT} from './prompts'
import {executeJavaScriptTool} from './tools'
import {CodeActState} from './types'
import * as dotenv from 'dotenv'

dotenv.config()

const llm = new ChatOpenAI({
  modelName: 'gpt-4o',
  apiKey: process.env.API_KEY || '',
  configuration: {
    baseURL: 'http://localhost:3001',
  },
})

// 提取代码块
function extractCode(content: string): string | null {
  if (content.includes('```javascript')) {
    const blocks = content.split('```javascript')
    if (blocks.length > 1 && blocks[1]) {
      const codeParts = blocks[1].split('```')
      if (codeParts[0]) {
        return codeParts[0].trim()
      }
    }
  }
  return null
}

// LLM 节点：分析问题并生成代码
async function llmCall(state: typeof CodeActState.State) {
  console.log('\n[LLM 节点] 分析问题...')

  const messages = [
    new SystemMessage(SYSTEM_PROMPT),
    new HumanMessage(state.userPrompt),
    ...state.messages,
  ]

  const response = await llm.invoke(messages)
  const code = extractCode(response.content as string)

  if (code) {
    console.log('[生成代码]')
    return {
      messages: [response],
      output: '',
      code,
    }
  }

  console.log('[直接给出答案]')
  return {
    messages: [response],
    output: response.content as string,
    code: null,
  }
}

// 路由函数：判断是否需要执行代码
function shouldExecute(state: typeof CodeActState.State): string {
  if (state.output) {
    return END
  }
  return 'executeNode'
}

// 执行节点：运行生成的代码
async function executeNode(state: typeof CodeActState.State) {
  console.log('\n[执行节点] 运行代码...')

  const code = state.code
  if (!code) {
    return {messages: []}
  }

  const result = await executeJavaScriptTool.invoke({code})

  return {
    messages: [new HumanMessage(`## 执行结果:\n${result}`)],
    code: null,
  }
}

// 构建 CodeAct 图
export function buildCodeActGraph() {
  const workflow = new StateGraph(CodeActState)
    .addNode('llmCall', llmCall)
    .addNode('executeNode', executeNode)
    .addEdge('__start__', 'llmCall')
    .addConditionalEdges('llmCall', shouldExecute, {
      executeNode: 'executeNode',
      [END]: END,
    })
    .addEdge('executeNode', 'llmCall')

  return workflow.compile()
}

// 运行 CodeAct Agent
export async function runCodeActAgent(query: string): Promise<string> {
  console.log('[启动 CodeAct Agent]')
  console.log(`[用户问题] ${query}\n`)

  const app = buildCodeActGraph()

  const result = await app.invoke({
    userPrompt: query,
  } as any)

  console.log('\n[完成]\n')
  return result.output
}
